### DISCUSSION TITLE
Is there a way to manually register and/or disable ConfigMappings?

### DISCUSSION THREAD

**Discussion by alex-martel** [quarkusio/quarkus#30970] [Q&A]
Hi all,
We have multiple services (beans), all implementing the same interface. For now we use `IfBuildProperty` to select which one will be used at runtime. Each build has it's own configuration file, added this way `quarkus.config.locations=${service}/${service}.properties`. These are mapped with `@ConfigMapping`.

I have few problems with this architecture, that I am not sure what would be the best way to solve.

1- `@IfBuildProperty(name = Service.property, stringValue = Service.Constants.FirstService)`
this seem to disable the dependent bean created, but still loads the config at startup, so when I switch to another service / another config this will fail the quarkus' startup, as the "first service"' config is not available, which is what I expect.

2- I then tried to not use `IfBuildProperty` to have everything modifiable at runtime, seems to work for my service as I can select via a Producer method. But I am not able to achieve the same for the ConfigMapping. I tried to remove the `@ConfigMapping(prefix = "prefix")` and  `@StaticInitSafe` from the class and register it myself in a producer method, something like this;
```java
@Produces
@Dependent
public ServiceConfig serviceConfig() {
    ConfigMappings.registerConfigMappings(ConfigProvider.getConfig().unwrap(SmallRyeConfig.class),
            Set.of(new ConfigMappings.ConfigClassWithPrefix(ServiceConfig.class, "service")));

return ConfigProvider.getConfig().unwrap(SmallRyeConfig.class).getConfigMapping(ServiceConfig.class);
```

But then the getConfigMapping fails, so I am unsure what would be the best way to achieve sort of optional ConfigMapping.

Also try using the `SmallRyeConfigBuilder` a bit like [this](https://stackoverflow.com/questions/74921793/getting-variables-from-application-properties-using-smallryeconfig-in-quarkus-fa) but was also unsuccessful

Help would be appreciated. 

Thanks,
Alex

**Comment by alex-martel** [comment]
I have some progress. So ultimately I was able to have that "optional" configmapping by doing the following;

1- Remove `@ConfigMapping` and `@StaticInitSafe` from the config mapping interface
2- I have a producer method that does the following before creating my service bean
```java
ConfigMappings.registerConfigMappings(ConfigProvider.getConfig().unwrap(SmallRyeConfig.class),
        Set.of(new ConfigMappings.ConfigClassWithPrefix(MyServiceConfig.class, "service")));
return new MyService();
```
3-I have another producer that creates the CDI dependent bean as `@ConfigMapping` seem to do, that way I can also Inject the manually registered / optional mapping.

```java
// Note not having the prefix in the getConfigMappin does not work, that was my initial mistake in the OP
return ConfigProvider.getConfig().unwrap(SmallRyeConfig.class).getConfigMapping(MyServiceConfig.class,"service")
```

The only problem I have now is that I am unable to achieve the bean validation my formerly `@ConfigMapping` was doing via annotation.

I tried doing it manually this way; but the validation always passes, I'd appreciate some pointers one how to get that one last piece work. cc @radcortez 
```java
// config is from SmallRyeConfig#getConfigMapping
ConfigValidator validator = new BeanValidationConfigValidatorImpl();
validator.validateMapping(MyServiceConfig.class, "service", config);
```

**Reply by radcortez** [reply]
I would not recommend using `ConfigMappings.registerConfigMappings`. This is ugly, it should not be a public API, but it is required to register mappings found via CDI at runtime (because we have no way no determine when `Config` is instantiated and it can be already available when the CDI container is bootstraping). This does not apply to Quarkus, because the discovery is done on build time and we can call the builder as expected.  This is required for other CDI implementations like Weld or OWB.

The validation does not work, because the validator is missing the classes to validate. These are registered at build time (we register all classes annotated with `@ConfigMapping`). Because this is build time, the validator does not introspect the instance to validate at runtime, so there are no validations and you don't get errors back.

Let's try to figure out a proper way for you to achieve this without having to workaround the code :)

**Comment by radcortez** [comment]
Are you applying the `@IfBuildProperty` to the `@ConfigMapping` class? Right now, we don't have support for it.

Some time ago, in https://github.com/quarkusio/quarkus/pull/19965, we no longer registered mappings injected in unused / removed beans. So if you have an injection in an unused bean the mapping should not be registered. I've tried this scenario in a sample project, and it seems to work, but I might be missing something.

I guess that your beans are annotated with `@IfBuildProfile`, and you only have injections to the required config classes. If other classes are still being registered, they are being caught in some other way and we need to understand why.

Can you send a reproducer to check the setup and investigate the problem?
